File: backend/app/Support/PointsSettings.php
<?php

namespace App\Support;

use App\Enums\AttendanceStatus;
use App\Enums\CurriculumType;
use App\Enums\RecitationGrade;
use App\Models\CurriculumItem;
use App\Models\Institute;

/**
 * إعدادات النقاط لمعهد، مقروءة من institutes.settings['points'] بقيم افتراضية.
 *
 * الغلاف يمنع تناثر data_get($institute->settings, 'points.…') في الإجراءات والشاشات،
 * ويجعل تغيير شكل الإعدادات مسألة ملف واحد.
 */
class PointsSettings
{
    /**
     * القيم الافتراضية — تُستعمل حين لا يضبط المعهد شيئاً، وتُملأ بها شاشة الإعدادات.
     *
     * @var array<string, mixed>
     */
    public const DEFAULTS = [
        'quran_per_15_lines' => 10,
        'hadith_per_item' => 5,
        'mutun_per_bayt' => 1,
        'attendance' => ['present' => 2, 'late' => 1, 'excused' => 0, 'absent' => 0],
        'grade_multiplier' => ['excellent' => 100, 'very_good' => 80, 'good' => 60],
    ];

    /**
     * @param  array<string, mixed>  $values
     */
    private function __construct(private readonly array $values) {}

    public static function for(?Institute $institute): self
    {
        return new self((array) data_get($institute?->settings, 'points', []));
    }

    /**
     * @param  array<string, mixed>  $values
     */
    public static function fromArray(array $values): self
    {
        return new self($values);
    }

    public function quranPer15Lines(): float
    {
        return (float) $this->value('quran_per_15_lines');
    }

    public function hadithPerItem(): float
    {
        return (float) $this->value('hadith_per_item');
    }

    public function mutunPerBayt(): float
    {
        return (float) $this->value('mutun_per_bayt');
    }

    public function attendance(AttendanceStatus $status): float
    {
        return (float) $this->value("attendance.{$status->value}");
    }

    /**
     * معامل التقدير كنسبة من الواحد: «جيد جداً» المخزَّن 80 يعود 0.80.
     */
    public function gradeMultiplier(?RecitationGrade $grade): float
    {
        if ($grade === null) {
            return 1.0;
        }

        return (float) $this->value("grade_multiplier.{$grade->value}") / 100;
    }

    /**
     * نقاط مدى قرآني من أسطره الجديدة — الصيغة الوحيدة في النظام.
     */
    public function quranPoints(float $newLines, ?RecitationGrade $grade): float
    {
        return round($newLines / 15 * $this->quranPer15Lines() * $this->gradeMultiplier($grade), 2);
    }

    /**
     * نقاط إنجاز بند منهج: النسبة × عدّاد البند × معامل النوع.
     *
     * البند بلا عدّاد (بند مخصّص، أو جزء قرآني) يُحسب وحدةً واحدة، فلا يضيع إنجازه
     * لمجرّد أن المشرف لم يملأ عدّاداً.
     */
    public function progressPoints(CurriculumItem $item, int $percent): float
    {
        $type = $item->curriculum?->type;

        [$count, $factor] = match ($type) {
            CurriculumType::Hadith => [(int) ($item->meta['hadiths'] ?? 1), $this->hadithPerItem()],
            CurriculumType::Mutun => [(int) ($item->meta['abyat'] ?? 1), $this->mutunPerBayt()],
            default => [0, 0.0],
        };

        return round($percent / 100 * max(1, $count) * $factor, 2);
    }

    /**
     * الإعدادات كاملةً بعد دمجها بالافتراضيات — تعبّئ بها شاشة المعهد حقولها.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'quran_per_15_lines' => $this->quranPer15Lines(),
            'hadith_per_item' => $this->hadithPerItem(),
            'mutun_per_bayt' => $this->mutunPerBayt(),
            'attendance' => collect(AttendanceStatus::cases())
                ->mapWithKeys(fn (AttendanceStatus $status) => [$status->value => $this->attendance($status)])
                ->all(),
            'grade_multiplier' => collect(RecitationGrade::cases())
                ->mapWithKeys(fn (RecitationGrade $grade) => [$grade->value => $this->gradeMultiplier($grade) * 100])
                ->all(),
        ];
    }

    private function value(string $key): mixed
    {
        return data_get($this->values, $key) ?? data_get(self::DEFAULTS, $key);
    }
}

File: packages/mousqe_core/lib/src/models/points_settings.dart
import 'enums.dart';

/// `institute.points` في `/bootstrap` — نظير `App\Support\PointsSettings::toArray()`،
/// ✅ م.6.6.
///
/// **ولماذا تصل في اللقطة أصلاً؟** لأن `institutes.settings` جدولٌ **لا يُزامَن**
/// ([SYNC-PROTOCOL.md §7](../../../../../docs/SYNC-PROTOCOL.md))، والعميلُ هو من
/// يحسب الإحصاءَ المشتقّ. فتقريرُ النقاط يجمع نقاطَ الحضور بقيمِ المعهد، وهي
/// قيمٌ لا تصل في أيّ تيّار.
///
/// وهذه **ثالثةُ مرّةٍ يتكرّر فيها النمط نفسُه** في هذا المشروع: ما يُحسب محلياً
/// يحتاج ثابتاً خادمياً، فيُحمَل في `/bootstrap` لا في جدولٍ يُزامَن — سبقتها
/// `theme` (م.5.2) و`attendance.late_grace_minutes` (م.5.3)، وهذه الأخيرةُ في
/// **نفس الحقل المجاور** حرفياً.
///
/// و**بلا freezed عمداً:** الحمولةُ خرائطُ متداخلة بمفاتيحَ نصّية (`attendance` و
/// `grade_multiplier`)، ونمذجتُها صفوفاً يعني صفَّين إضافيَّين لا يُقرآن إلا من
/// هنا. والافتراضياتُ هي **نفسُ `PointsSettings::DEFAULTS`**، فخادمٌ أقدم من م.6.6
/// لا يبعث الحقل يبقى عاملاً بها لا بأصفار.
class PointsSettings {
  const PointsSettings({
    this.quranPer15Lines = 10,
    this.hadithPerItem = 5,
    this.mutunPerBayt = 1,
    this.presentPoints = 2,
    this.latePoints = 1,
    this.excusedPoints = 0,
    this.absentPoints = 0,
  });

  final double quranPer15Lines;
  final double hadithPerItem;
  final double mutunPerBayt;

  final double presentPoints;
  final double latePoints;
  final double excusedPoints;
  final double absentPoints;

  /// نقاطُ صفِّ حضورٍ واحد بحالته.
  double attendance(AttendanceStatus status) => switch (status) {
        AttendanceStatus.present => presentPoints,
        AttendanceStatus.late => latePoints,
        AttendanceStatus.excused => excusedPoints,
        AttendanceStatus.absent => absentPoints,
      };

  factory PointsSettings.fromJson(Map<String, dynamic> json) {
    final attendance = (json['attendance'] as Map?)?.cast<String, dynamic>();

    const fallback = PointsSettings();

    double read(dynamic value, double orElse) =>
        value is num ? value.toDouble() : orElse;

    return PointsSettings(
      quranPer15Lines:
          read(json['quran_per_15_lines'], fallback.quranPer15Lines),
      hadithPerItem: read(json['hadith_per_item'], fallback.hadithPerItem),
      mutunPerBayt: read(json['mutun_per_bayt'], fallback.mutunPerBayt),
      presentPoints:
          read(attendance?['present'], fallback.presentPoints),
      latePoints: read(attendance?['late'], fallback.latePoints),
      excusedPoints: read(attendance?['excused'], fallback.excusedPoints),
      absentPoints: read(attendance?['absent'], fallback.absentPoints),
    );
  }

  Map<String, dynamic> toJson() => {
        'quran_per_15_lines': quranPer15Lines,
        'hadith_per_item': hadithPerItem,
        'mutun_per_bayt': mutunPerBayt,
        'attendance': {
          'present': presentPoints,
          'late': latePoints,
          'excused': excusedPoints,
          'absent': absentPoints,
        },
      };
}
